src/datafusion/helpers.rs

```diff
@@ -0,0 +1,228 @@
+        | Expr::WindowFunction { .. }
+        | Expr::Wildcard { .. }
+        | Expr::Unnest { .. }
+        | Expr::Placeholder(_) => {
+            is_applicable = false;
+            Ok(VisitRecursion::Stop)
+        }
+    })
+    .unwrap();
+    is_applicable
+}
+
+// Below is all the logic necessary (I think) to convert a PhysicalExpr into a ZarrChunkFilter.
+// The logic is mostly copied from datafusion, and is simplified here for the zarr use case.
+pub struct ZarrFilterCandidate {
+    expr: Arc<dyn PhysicalExpr>,
+    projection: Vec<usize>,
+}
+
+struct ZarrFilterCandidateBuilder<'a> {
+    expr: Arc<dyn PhysicalExpr>,
+    file_schema: &'a Schema,
+    required_column_indices: BTreeSet<usize>,
+}
+
+impl<'a> ZarrFilterCandidateBuilder<'a> {
+    pub fn new(expr: Arc<dyn PhysicalExpr>, file_schema: &'a Schema) -> Self {
+        Self {
+            expr,
+            file_schema,
+            required_column_indices: BTreeSet::default(),
+        }
+    }
+
+    pub fn build(mut self) -> DataFusionResult<Option<ZarrFilterCandidate>> {
+        let expr = self.expr.clone().rewrite(&mut self)?;
+
+        Ok(Some(ZarrFilterCandidate {
+            expr,
+            projection: self.required_column_indices.into_iter().collect(),
+        }))
+    }
+}
+
+impl<'a> TreeNodeRewriter for ZarrFilterCandidateBuilder<'a> {
+    type N = Arc<dyn PhysicalExpr>;
+
+    fn pre_visit(&mut self, node: &Arc<dyn PhysicalExpr>) -> DataFusionResult<RewriteRecursion> {
+        if let Some(column) = node.as_any().downcast_ref::<Column>() {
+            if let Ok(idx) = self.file_schema.index_of(column.name()) {
+                self.required_column_indices.insert(idx);
+            }
+        }
+
+        Ok(RewriteRecursion::Continue)
+    }
+
+    fn mutate(&mut self, expr: Arc<dyn PhysicalExpr>) -> DataFusionResult<Arc<dyn PhysicalExpr>> {
+        Ok(expr)
+    }
+}
+
+#[derive(Clone)]
+pub struct ZarrDatafusionArrowPredicate {
+    physical_expr: Arc<dyn PhysicalExpr>,
+    projection_mask: ZarrProjection,
+    projection: Vec<String>,
+}
+
+impl ZarrDatafusionArrowPredicate {
+    pub fn new(candidate: ZarrFilterCandidate, schema: &Schema) -> DataFusionResult<Self> {
+        let cols: Vec<_> = candidate
+            .projection
+            .iter()
+            .map(|idx| schema.field(*idx).name().to_string())
+            .collect();
+
+        let schema = Arc::new(schema.project(&candidate.projection)?);
+        let physical_expr = reassign_predicate_columns(candidate.expr, &schema, true)?;
+
+        Ok(Self {
+            physical_expr,
+            projection_mask: ZarrProjection::keep(cols.clone()),
+            projection: cols,
+        })
+    }
+}
+
+impl ZarrArrowPredicate for ZarrDatafusionArrowPredicate {
+    fn projection(&self) -> &ZarrProjection {
+        &self.projection_mask
+    }
+
+    fn evaluate(&mut self, batch: &RecordBatch) -> Result<BooleanArray, ArrowError> {
+        let index_projection = self
```

Comment:
This is the bit that depends on the column names. Here, the incoming record batch can have any number of columns, it doesn't matter, as long as it contains at least the columns the predicate needs. In the parquet implementation, again if I understood correctly, it's expected to come in with only the required columns, but by using column names here, that's not required, we figure out the indices in the record batch based on the column names, and re-project it before passing it to the physical expression. The re-projection does still happen in the parquet implementation, I think to handle different column orderings, but here we use it to also drop unnecessary columns, that way, for example if the predicate only requires the `lon` column, we can re-use a record batch that contains `lat, lon`.